### PR TITLE
[FEATURE] Prearm

### PR DESCRIPTION
**Pull request by harryjph**
This PR implements a Prearm (#5595). It adds a new mode switch ~~which will need a relevant configurator PR.~~

~~I also took the opportunity to fix [this FIXME](https://github.com/iNavFlight/inav/blob/b322940dd78f542864a0d091b5d9420d35c5dd18/src/main/fc/runtime_config.h#L20) by separating out the `armingFlags` enum into Arming Flags (flags relating to the current arming) and Arming Disable Flags (flags that prevent arming). This forms the bulk of this PR as I gave the macros clearer names.~~

### CODE REVIEW

**Review by DzikuVx — APPROVED** [review]
LGTM @digitalentity WDYT?

File: src/main/fc/fc_msp.c [hunk 1]
```diff
@@ -452,7 +452,7 @@ static bool mspFcProcessOutCommand(uint16_t cmdMSP, sbuf_t *dst, mspPostProcessF
 
             if (cmdMSP == MSP_STATUS_EX) {
                 sbufWriteU16(dst, averageSystemLoadPercent);
-                sbufWriteU16(dst, armingFlags);
+                sbufWriteU16(dst, armingDisableFlags);
```
Comment (digitalentity) [inline src/main/fc/fc_msp.c]: This is a breaking change. The `ARMED` bit of this field is used by external devices to detect whether the aircraft is in the armed state.

File: src/main/fc/fc_core.c [hunk 2]
```diff
@@ -194,115 +194,122 @@ static void updateArmingStatus(void)
...
-        else {
-	       DISABLE_ARMING_FLAG(ARMING_DISABLED_SERVO_AUTOTRIM);
-	    }
+            SET_ARMING_DISABLED_FLAG(ARMING_DISABLED_SERVO_AUTOTRIM);
+        } else {
+            UNSET_ARMING_DISABLED_FLAG(ARMING_DISABLED_SERVO_AUTOTRIM);
+        }
+
+        if (isModeActivationConditionPresent(BOXPREARM)) {
+            if (IS_RC_MODE_ACTIVE(BOXPREARM) && !ARMING_FLAG(WAS_ARMED_WITH_PREARM)) {
```
Comment (digitalentity) [inline src/main/fc/fc_core.c]: Why additional `WAS_ARMED_WITH_PREARM` flag? IMO the `ARMING_DISABLED_NO_PREARM` flag can only be changed in the "disarmed" state.

File: src/main/fc/fc_core.c [hunk 3]
```diff
@@ -194,115 +194,122 @@ static void updateArmingStatus(void)
...
-	    }
-        else {
-	       DISABLE_ARMING_FLAG(ARMING_DISABLED_SERVO_AUTOTRIM);
-	    }
+            SET_ARMING_DISABLED_FLAG(ARMING_DISABLED_SERVO_AUTOTRIM);
+        } else {
+            UNSET_ARMING_DISABLED_FLAG(ARMING_DISABLED_SERVO_AUTOTRIM);
+        }
+
+        if (isModeActivationConditionPresent(BOXPREARM)) {
```
Comment (digitalentity) [inline src/main/fc/fc_core.c:298]: `isModeActivationConditionPresent` is expensive in terms of CPU time. We should process it once and expose it as a special function call reading a boolean flag, similar to checking for navigation modes.

**Review by digitalentity — CHANGES_REQUESTED** [review]
Looks good in general. Just a few things:

1. Please remove the breaking separation into `armingFlags` and `armingDisabledFlags`. Configurator and other external peripherals depend on that bitfield.
2. I don't fully undestand the logic behind the additional `ARMED_WITH_PREARM` flag. Can't we process the prearm bit always?
3. Optimization regarding checking for "prearm" mode being present. We shouldn't be evaluating the whole aux table every cycle.

File: src/main/fc/fc_msp.c (hunk 1 above)
Comment (digitalentity) [reply src/main/fc/fc_msp.c]: I suggest the FIXME should be forked out of this PR into a separate one for further discussion.

**Comment by harry1453** [comment]
Okay, do you think that setting the Arm disable flags back to their original bit positions will resolve the breaking change I.E. does configurator etc. use the ARMED and WAS_EVER_ARMED flags? Or should I remove the whole thing?

**Comment by stronnag** [comment]
Please do not break established APIs; they are used by multiple 3rd party applications.

**Comment by harry1453** [comment]
Thanks for the review! In response to the review:

1. Done
2. Honestly can't remember myself, I've removed it now.
3. Done, please see second commit

File: src/main/fc/fc_core.c [hunk 4]
```diff
@@ -295,6 +295,14 @@ static void updateArmingStatus(void)
 	       DISABLE_ARMING_FLAG(ARMING_DISABLED_SERVO_AUTOTRIM);
 	    }
 
+        if (isModeActivationConditionPresent(BOXPREARM)) {
+            if (IS_RC_MODE_ACTIVE(BOXPREARM)) {
+                DISABLE_ARMING_FLAG(ARMING_DISABLED_NO_PREARM);
```
Comment (digitalentity) [inline src/main/fc/fc_core.c:300]: We should also disable `ARMING_DISABLED_NO_PREARM` if `BOXPREARM` is not configured. This way we won't end up in a situation where a person configuring the aircraft will get stuck being unable to arm if they configure and then remove the `PREARM` mode.

File: src/main/fc/rc_modes.c [hunk 5]
```diff
@@ -134,13 +134,7 @@ void rcModeUpdate(boxBitmask_t *newState)
...
 bool isModeActivationConditionPresent(boxId_e modeId)
 {
-    for (int index = 0; index < MAX_MODE_ACTIVATION_CONDITION_COUNT; index++) {
-        if (modeActivationConditions(index)->modeId == modeId && IS_RANGE_USABLE(&modeActivationConditions(index)->range)) {
-            return true;
-        }
-    }
-
-    return false;
+    return specifiedConditionCountPerMode[modeId] > 0;
```
Comment (digitalentity) [inline src/main/fc/rc_modes.c:137]: Nice find! I forgot that we actually do pre-calculate this :+1:

**Review by digitalentity — CHANGES_REQUESTED** [review]
A minor nit to cover for an edge-case. Otherwise - LGTM

File: src/main/fc/rc_modes.c (hunk 5 above)
Comment (harryjph) [reply src/main/fc/rc_modes.c:137]: I made a system that used a bitflag and as I was searching for places that should trigger it updating, I found that this does the same thing :)

File: src/main/fc/fc_core.c (hunk 4 above)
Comment (harryjph) [reply src/main/fc/fc_core.c:300]: Good catch! Didn't think of that.

**Review by digitalentity — APPROVED** [review]

**Pull request merged**